File: packages/runtime-web-three/src/bevyCatalogResiduals.ts
import { diagnoseBevyCatalogResidualDeclarations, targetProfileOutputDiagnostic, type BevyCatalogTargetProfileOutput, type IIrDiagnostic, type IRuntimeDiagnostic, type IWorldEntity, type IWorldIr } from "@threenative/ir";

export interface IWebQueryCombinationObservation {
  a: string;
  b: string;
  order: number;
}

export interface IWebDisabledEntityQueryParticipationReport {
  entity: string;
  participatesInQueries: boolean;
  policy: "portable-participation-state";
  rendererVisibility: "unchanged";
  schema: "threenative.bevy-catalog.ecs";
  version: "0.1.0";
}

export interface IWebTextInputEvent {
  action: "commit" | "input";
  order: number;
  value: string;
}

export interface IWebWindowPolicyReport {
  diagnostics: IRuntimeDiagnostic[];
  resize: {
    height: number;
    scaleFactor: number;
    width: number;
  };
  schema: "threenative.bevy-catalog.window";
  version: "0.1.0";
}

export interface IWebGeneratedAssetPolicyReport {
  assetId: string;
  path: string;
  schema: string;
  status: "bundle-artifact";
}

export interface IWebGltfMetadataTransformPolicyReport {
  extension: string;
  processor: "metadata";
  schema: "threenative.bevy-catalog.assets.gltf-metadata-transform";
  transform: "AnimationGraph";
  version: "0.1.0";
}

export function traceWebQueryCombinations(world: IWorldIr, component: string, limit = Number.POSITIVE_INFINITY): IWebQueryCombinationObservation[] {
  const entities = world.entities
    .filter((entity) => hasComponent(entity, component))
    .map((entity) => entity.id)
    .sort();
  const observations: IWebQueryCombinationObservation[] = [];
  for (let left = 0; left < entities.length; left += 1) {
    for (let right = left + 1; right < entities.length; right += 1) {
      if (observations.length >= limit) {
        return observations;
      }
      observations.push({ a: entities[left] ?? "", b: entities[right] ?? "", order: observations.length + 1 });
    }
  }
  return observations;
}

export function reportWebDisabledEntityQueryParticipation(entity: string, participatesInQueries: boolean): IWebDisabledEntityQueryParticipationReport {
  return {
    entity,
    participatesInQueries,
    policy: "portable-participation-state",
    rendererVisibility: "unchanged",
    schema: "threenative.bevy-catalog.ecs",
    version: "0.1.0",
  };
}

export function traceWebTextInputEvents(values: readonly string[]): IWebTextInputEvent[] {
  return values.map((value, index) => ({
    action: index === values.length - 1 ? "commit" : "input",
    order: index + 1,
    value,
  }));
}

export function reportWebWindowCatalogPolicy(width: number, height: number, scaleFactor: number): IWebWindowPolicyReport {
  const diagnostics = diagnoseBevyCatalogResidualDeclarations({
    uiWindow: {
      windowPolicy: {
        clearColorRuntimeUpdate: true,
        cursorImage: "assets/cursor.png",
        lowPowerPresentMode: true,
        multiWindow: true,
      },
    },
  }).map((diagnostic) => ({ ...diagnostic, severity: diagnostic.severity ?? "error" } as IRuntimeDiagnostic));
  return {
    diagnostics,
    resize: { height, scaleFactor, width },
    schema: "threenative.bevy-catalog.window",
    version: "0.1.0",
  };
}

export function reportWebGeneratedAssetPolicy(assetId: string, schema: string): IWebGeneratedAssetPolicyReport {
  return {
    assetId,
    path: `artifacts/generated/${assetId}.json`,
    schema,
    status: "bundle-artifact",
  };
}

export function reportWebGltfMetadataTransformPolicy(extension: string, transform: "AnimationGraph"): IWebGltfMetadataTransformPolicyReport {
  return {
    extension,
    processor: "metadata",
    schema: "threenative.bevy-catalog.assets.gltf-metadata-transform",
    transform,
    version: "0.1.0",
  };
}

export function reportWebTargetProfileOutputDiagnostic(output: BevyCatalogTargetProfileOutput, targets: readonly string[], path = "target.profile.json/targets"): IIrDiagnostic {
  return targetProfileOutputDiagnostic(output, output === "web" ? "web" : "desktop", targets, path);
}

function hasComponent(entity: IWorldEntity, component: string): boolean {
  return Object.prototype.hasOwnProperty.call(entity.components, component);
}

File: runtime-bevy/crates/threenative_runtime/src/bevy_catalog_residuals.rs
use serde_json::{Value, json};

pub fn report_disabled_entity_query_participation(entity: &str, participates: bool) -> Value {
    json!({
        "entity": entity,
        "participatesInQueries": participates,
        "policy": "portable-participation-state",
        "rendererVisibility": "unchanged",
        "schema": "threenative.bevy-catalog.ecs",
        "version": "0.1.0"
    })
}

pub fn report_query_combinations(mut entities: Vec<&str>, limit: usize) -> Value {
    entities.sort_unstable();
    let mut observations = Vec::new();
    'outer: for left in 0..entities.len() {
        for right in (left + 1)..entities.len() {
            if observations.len() >= limit {
                break 'outer;
            }
            observations.push(json!({
                "a": entities[left],
                "b": entities[right],
                "order": observations.len() + 1
            }));
        }
    }
    json!({
        "observations": observations,
        "schema": "threenative.bevy-catalog.ecs.query-combinations",
        "version": "0.1.0"
    })
}

pub fn report_window_resize_and_scale_factor(width: u32, height: u32, scale_factor: f64) -> Value {
    json!({
        "diagnostics": [
            {
                "code": "TN_CATALOG_WINDOW_CURSOR_UNSUPPORTED",
                "message": "Custom cursor images and cursor animation are diagnostic-only until a portable cursor policy exists.",
                "path": "bevy-catalog-residuals.json/uiWindow/windowPolicy/cursorImage",
                "severity": "error",
                "suggestion": "Use the portable primary-window runtime configuration or a target-specific adapter."
            },
            {
                "code": "TN_CATALOG_WINDOW_POWER_POLICY_UNSUPPORTED",
                "message": "Low-power present mode and background throttling are host policies outside the portable runtime contract.",
                "path": "bevy-catalog-residuals.json/uiWindow/windowPolicy/lowPowerPresentMode",
                "severity": "error",
                "suggestion": "Use the portable primary-window runtime configuration or a target-specific adapter."
            },
            {
                "code": "TN_CATALOG_WINDOW_CLEAR_COLOR_RUNTIME_UNSUPPORTED",
                "message": "Runtime clear-color/window background mutation is not promoted without shared web/native observation evidence.",
                "path": "bevy-catalog-residuals.json/uiWindow/windowPolicy/clearColorRuntimeUpdate",
                "severity": "error",
                "suggestion": "Use the portable primary-window runtime configuration or a target-specific adapter."
            },
            {
                "code": "TN_CATALOG_WINDOW_MULTI_WINDOW_UNSUPPORTED",
                "message": "Portable runtime bundles are single-window; per-window targets remain diagnostic-only.",
                "path": "bevy-catalog-residuals.json/uiWindow/windowPolicy/multiWindow",
                "severity": "error",
                "suggestion": "Use the portable primary-window runtime configuration or a target-specific adapter."
            }
        ],
        "resize": {
            "height": height,
            "scaleFactor": scale_factor,
            "width": width
        },
        "schema": "threenative.bevy-catalog.window",
        "version": "0.1.0"
    })
}

pub fn unsupported_gltf_executable_extension_processor(extension: &str, path: &str) -> Value {
    json!({
        "code": "TN_CATALOG_GLTF_EXTENSION_PROCESSOR_UNSUPPORTED",
        "message": format!("glTF extension '{extension}' uses an executable processor outside the portable import policy."),
        "path": path,
        "severity": "error",
        "suggestion": "Use schema-backed metadata transforms such as declared AnimationGraph import metadata."
    })
}

pub fn report_gltf_metadata_transform_policy(extension: &str, transform: &str) -> Value {
    json!({
        "extension": extension,
        "processor": "metadata",
        "schema": "threenative.bevy-catalog.assets.gltf-metadata-transform",
        "transform": transform,
        "version": "0.1.0"
    })
}

pub fn target_profile_output_diagnostic(output: &str, targets: Vec<&str>, path: &str) -> Value {
    let required_target = if output == "web" { "web" } else { "desktop" };
    let suggestion = if required_target == "web" {
        "Add 'web' to target.profile.json targets or choose a non-web output."
    } else {
        "Add 'desktop' to target.profile.json targets for offline, native, or package outputs."
    };
    json!({
        "code": "TN_CATALOG_TARGET_PROFILE_OUTPUT_UNSUPPORTED",
        "message": format!("Target profile for '{output}' output must include '{required_target}'."),
        "path": path,
        "severity": "error",
        "suggestion": suggestion,
        "target": output,
        "value": targets
    })
}

pub fn report_generated_asset_policy(asset_id: &str, schema: &str) -> Value {
    json!({
        "assetId": asset_id,
        "path": format!("artifacts/generated/{asset_id}.json"),
        "schema": schema,
        "status": "bundle-artifact"
    })
}
